pkg/controller/backup/backup_controller.go

```diff
@@ -156,8 +158,34 @@ func (c *Controller) updateBackup(cur interface{}) {
 		return
 	}
 
-	if v1alpha1.IsBackupScheduled(newBackup) {
-		klog.V(4).Infof("backup %s/%s is already scheduled, skipping", ns, name)
+	if v1alpha1.IsBackupFailed(newBackup) {
+		klog.V(4).Infof("backup %s/%s is Failed, skipping.", ns, name)
+		return
+	}
+
+	if v1alpha1.IsBackupScheduled(newBackup) || v1alpha1.IsBackupRunning(newBackup) || v1alpha1.IsBackupPrepare(newBackup) {
+		selector, err := label.NewBackup().Instance(newBackup.GetInstanceName()).BackupJob().Backup(name).Selector()
```

Comment:
this code block is too nested, do you have some ideas on making it simpler?

Reply:
I previously want to `return` as early as possible, but it seems still ugly. any suggestion?

Reply:
maybe something like this, not tested
```go
if v1alpha1.IsBackupScheduled(newBackup) || v1alpha1.IsBackupRunning(newBackup) || v1alpha1.IsBackupPrepare(newBackup) {
	// check for failed backup pods, and update the condition
	var selector labels.Selector
	var pods []*v1.Pod
	klog.V(4).Infof("backup %s/%s is already Scheduled, Running, Preparing or Failed, skipping.", ns, name)
	if selector, err := label.NewBackup().Instance(newBackup.GetInstanceName()).BackupJob().Backup(name).Selector(); err != nil {
		// log err
		return
	}
	if pods, err := c.deps.PodLister.Pods(ns).List(selector); err != nil {
		// log err
		return
	}
	for _, pod := range pods {
		if pod.Status.Phase == corev1.PodFailed {
			klog.Infof("backup %s/%s has failed pod %s.", ns, name, pod.Name)
			if err := c.control.UpdateCondition(newBackup, &v1alpha1.BackupCondition{
				Type:    v1alpha1.BackupFailed,
				Status:  corev1.ConditionTrue,
				Reason:  "AlreadyFailed",
				Message: fmt.Sprintf("Pod %s has failed", pod.Name),
			}); err != nil {
				klog.Errorf("Fail to update the condition of backup %s/%s", ns, name)
			}
			break
		}
	}
	return
}
```

Reply:
updated in https://github.com/pingcap/tidb-operator/pull/3696/commits/bdf60cc4106815e9b19f4204ba56b67ffaf5a2ab.